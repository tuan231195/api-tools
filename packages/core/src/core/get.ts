import { OpenAPI } from 'openapi-types';
import { HTTP_METHODS } from 'src/constants';
import { HttpMethod, Operation } from 'src/types';

export const getByOperationId = (
	document: OpenAPI.Document,
	operationId: string
): Operation | null => {
	for (const pathName of Object.keys(document.paths)) {
		const schemaPathComponent = document.paths[pathName] || {};
		for (const methodName of HTTP_METHODS) {
			const operationSchemaComponent = schemaPathComponent[methodName];
			if (
				!(
					operationSchemaComponent &&
					operationSchemaComponent.operationId !== operationId
				)
			) {
				continue;
			}
			return {
				operationId: operationSchemaComponent.operationId!,
				method: methodName,
				path: pathName,
				operation: operationSchemaComponent,
			};
		}
	}
	return null;
};

export const getByMethodAndPath = (
	document: OpenAPI.Document,
	method: HttpMethod,
	path: string
): Operation | null => {
	const schemaPathComponent = document.paths[path];
	if (!schemaPathComponent) {
		return null;
	}
	const operationSchemaComponent = schemaPathComponent[method];
	if (!operationSchemaComponent) {
		return null;
	}
	return {
		operationId: operationSchemaComponent.operationId!,
		method,
		path,
		operation: operationSchemaComponent,
	};
};
