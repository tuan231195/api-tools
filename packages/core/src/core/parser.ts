import SwaggerParser from '@apidevtools/swagger-parser';
import { parseSchema } from 'dtsgenerator';
import { omit, omitBy, startCase } from 'lodash';
import { OpenAPI } from 'openapi-types';
import { HTTP_METHODS, JSON_SCHEMA_7 } from 'src/constants';
import logger from 'src/logger';
import {
	JsonSchema,
	Parameter,
	ParsedSchema,
	ParsedSchemaOperation,
	ParsedSchemaRequestBody,
	ParsedSchemaResponseBody,
	Schema,
	SchemaObject,
	supportedMediaTypes,
} from 'src/types';

export async function validate(uri: string): Promise<OpenAPI.Document> {
	return SwaggerParser.validate(uri);
}

export const parse = async (
	uri: string | OpenAPI.Document
): Promise<ParsedSchema> => {
	const schema = typeof uri === 'string' ? await validate(uri) : uri;
	const result: ParsedSchema = {};

	await Promise.all(
		Object.keys(schema.paths).map(async (pathName) => {
			const schemaPathComponent = schema.paths[pathName] || {};
			for (const methodName of HTTP_METHODS) {
				const schemaOperationComponent =
					schemaPathComponent[methodName];
				if (!schemaOperationComponent) {
					continue;
				}
				if (!schemaOperationComponent.operationId) {
					logger(
						`Skip ${methodName.toUpperCase()} ${pathName}. Operation does not have operationId`
					);
					continue;
				}
				const defaultParameters = schemaPathComponent['parameters'];
				result[schemaOperationComponent.operationId] = {
					schema: await generateApiPathSchema(
						schemaOperationComponent,
						defaultParameters
					),
					operation: {
						operationId: schemaOperationComponent.operationId,
						operation: schemaOperationComponent,
						path: pathName,
						method: methodName,
					},
				};
			}
		})
	);

	return result;
};

const generateApiPathSchema = async (
	schema: OpenAPI.Operation,
	defaultParameters: OpenAPI.Parameters = []
): Promise<ParsedSchemaOperation> => {
	logger(`Processing ${schema.operationId}`);

	const operationParameters = schema.parameters ?? defaultParameters;
	const parametersSchema = generateParametersSchema(
		operationParameters as Parameter[]
	);
	const requestBodySchema = generateRequestBodySchema(schema);
	const responseBodySchema = generateResponseBodySchema(schema);
	return {
		headers: parametersSchema['header'] ?? null,
		queries: parametersSchema['query'] ?? null,
		params: parametersSchema['path'] ?? null,
		requestBody: requestBodySchema,
		responseBody: responseBodySchema,
	};
};

const generateRequestBodySchema = (
	schema: OpenAPI.Operation
): ParsedSchemaRequestBody | null => {
	let requestBody = getRequestBodySchema(schema);
	if (requestBody) {
		return requestBody;
	}
	requestBody = getBodyParameter(schema);
	return requestBody;
};

const generateResponseBodySchema = (
	schemaOperation: OpenAPI.Operation
): ParsedSchemaResponseBody => {
	if (!schemaOperation.responses) {
		return {
			statuses: {},
			all: null,
			success: null,
		};
	}
	const allStatusSchema = getAllStatusSchema(schemaOperation);
	const allSchema: Schema = toSchema({
		$id: 'Response',
		oneOf: Object.values(allStatusSchema).map((response) => ({
			$ref: response.content.$id,
		})),
	});

	const successResponses: Schema[] = Object.entries(allStatusSchema)
		.filter(([status]) => {
			const statusCode = parseInt(status, 10);
			return statusCode >= 200 && statusCode < 300;
		})
		.map(([, schema]) => schema);

	let successSchema: Schema | null = null;
	if (successResponses.length > 0) {
		successSchema = toSchema({
			$id: 'ResponseOk',
			oneOf: successResponses.map((response) => {
				return {
					$ref: response.content.$id,
				};
			}),
		});
	}

	return {
		all: allSchema,
		success: successSchema,
		statuses: allStatusSchema,
	};
};

export const getAllStatusSchema = (
	schemaOperation: OpenAPI.Operation
): ParsedSchemaResponseBody['statuses'] => {
	return Object.entries(schemaOperation.responses || []).reduce(
		(agg, [status, response]) => {
			const statusCode =
				status === 'default' ? 200 : parseInt(status, 10);
			const id = `Response${statusCode}`;
			const jsonBody =
				response.content && response.content['application/json'];

			const schema = jsonBody
				? toSchema({
						...jsonBody.schema,
						$id: id,
						example: jsonBody.example ?? null,
				  })
				: nullSchema(id);
			return {
				...agg,
				[statusCode]: schema,
			};
		},
		{} as ParsedSchemaResponseBody['statuses']
	);
};

const getRequestBodySchema = (
	schemaOperation: OpenAPI.Operation
): ParsedSchemaRequestBody | null => {
	const content = (schemaOperation as any).requestBody?.content;
	const requestContentType = supportedMediaTypes.find(
		(supportedType) => content && content[supportedType]
	);
	const requestBody = content && content[requestContentType || ''];
	if (!(requestBody && requestContentType)) {
		return null;
	}
	return {
		type: requestContentType === 'application/json' ? 'json' : 'form',
		contentType: requestContentType,
		schema: toSchema({
			...requestBody.schema,
			$id: 'RequestBody',
			example: requestBody.example ?? null,
		}),
	};
};

const getBodyParameter = (
	schemaOperation: OpenAPI.Operation
): ParsedSchemaRequestBody | null => {
	const parameterGroup = getParameterGroup(
		(schemaOperation.parameters || []) as Parameter[]
	);

	let schema: Schema | null = null;
	let requestType: ParsedSchemaRequestBody['type'] | null = null;
	let requestContentType: ParsedSchemaRequestBody['contentType'] | null =
		null;
	if (parameterGroup['body'] && parameterGroup['body'].length > 0) {
		requestType = 'json';
		requestContentType = 'application/json';
		const bodyParameter = parameterGroup['body'][0];
		schema = toSchema({
			...bodyParameter.schema,
			$id: 'RequestBody',
			example: bodyParameter.example ?? null,
		});
	} else if (parameterGroup['formData']) {
		requestType = 'form';
		requestContentType =
			'consumes' in schemaOperation &&
			(schemaOperation.consumes || []).includes('multipart/form-data')
				? 'multipart/form-data'
				: 'application/x-www-form-urlencoded';
		const bodyParameters = parameterGroup['formData'];
		schema = parseParameters('RequestBody', bodyParameters);
	}

	if (!(schema && requestType && requestContentType)) {
		return null;
	}

	return {
		type: requestType,
		contentType: requestContentType,
		schema,
	};
};

const getParameterGroup = (parameters: Parameter[]) => {
	const parameterGroup: Record<string, Parameter[]> = {};
	for (const parameter of parameters) {
		const schemaType = parameter.in;
		(parameterGroup[schemaType] = parameterGroup[schemaType] || []).push(
			parameter
		);
	}
	return parameterGroup;
};

const generateParametersSchema = (parameters: Parameter[]) => {
	const parameterGroup = getParameterGroup(parameters);
	// we don't want body parameters
	delete parameterGroup['body'];
	delete parameterGroup['formData'];

	return Object.entries(parameterGroup).reduce((agg, [group, parameters]) => {
		const id = startCase(group);
		return {
			...agg,
			[group]: parseParameters(id, parameters),
		};
	}, {} as Record<string, Schema>);
};

const parseParameters = (id: string, parameters: Parameter[]) => {
	const parameterProperties = parameters.reduce((agg, parameter) => {
		const parameterSchema =
			'schema' in parameter
				? parameter.schema
				: omit(parameter, ['collectionFormat', 'in', 'name']);

		return {
			...agg,
			[parameter.name]: parameterSchema,
		};
	}, {} as Record<string, SchemaObject>);

	const requiredProps = parameters
		.filter((parameter) => parameter.required)
		.map((parameter) => parameter.name);

	return toSchema({
		type: 'object',
		$id: id,
		properties: parameterProperties as any,
		additionalProperties: false,
		required: requiredProps,
	});
};

const nullSchema = (id: string) => {
	return toSchema({
		$id: id,
		type: 'boolean',
		not: true,
	});
};

const toSchema = (schema: JsonSchema): Schema => {
	return parseSchema({
		...omitBy(schema, (value) => value == null),
		$schema: JSON_SCHEMA_7,
	}) as Schema;
};
