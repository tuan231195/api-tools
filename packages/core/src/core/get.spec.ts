import path from 'path';
import { parse, validate } from './parser';
import { getByMethodAndPath, getByOperationId, getOperation } from './get';

const testDir = path.resolve(__dirname, '..', '..', '..', '..', 'test');

describe('getByOperationId', () => {
	it.each([
		['/v2/petstore.json', path.resolve(testDir, 'v2', 'petstore.json')],
		['/v3/petstore.json', path.resolve(testDir, 'v3', 'petstore.json')],
	])('%s', async (testName, filePath) => {
		const document = await validate(filePath);
		let operation = await getByOperationId(document, 'addPet');
		expect(operation).toMatchSnapshot();

		operation = await getByOperationId(document, 'blah');
		expect(operation).toBeNull();
	});
});

describe('getByMethodAndPath', () => {
	it.each([
		['/v2/petstore.json', path.resolve(testDir, 'v2', 'petstore.json')],
		['/v3/petstore.json', path.resolve(testDir, 'v3', 'petstore.json')],
	])('%s', async (testName, filePath) => {
		const document = await validate(filePath);
		let operation = await getByMethodAndPath(document, 'post', '/pet');
		expect(operation).toMatchSnapshot();

		operation = await getByMethodAndPath(document, 'post', '/blah');
		expect(operation).toBeNull();
	});
});

describe('getByMethodAndPath', () => {
	it.each([
		['/v2/petstore.json', path.resolve(testDir, 'v2', 'petstore.json')],
		['/v3/petstore.json', path.resolve(testDir, 'v3', 'petstore.json')],
	])('%s', async (testName, filePath) => {
		const document = await parse(filePath);
		let operation = await getOperation(document, { operationId: 'addPet' });
		expect(operation).toMatchSnapshot();

		operation = await getOperation(document, {
			methodName: 'post',
			path: '/pet',
		});
		expect(operation).toMatchSnapshot();
	});
});
