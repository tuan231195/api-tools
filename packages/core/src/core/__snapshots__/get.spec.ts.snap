// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getByMethodAndPath /v2/petstore.json 1`] = `{"operationId":"addPet","method":"post","path":"/pet"}`;

exports[`getByMethodAndPath /v2/petstore.json 2`] = `
Object {
  "info": {"operationId":"addPet","path":"/pet","method":"post"},
  "schema": Object {
    "headers": null,
    "params": null,
    "queries": null,
    "requestBody": Object {
      "contentType": "application/json",
      "schema": {"type":"object","required":["name","photoUrls"],"properties":{"id":{"type":"integer","format":"int64"},"category":{"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}},"xml":{"name":"Category"}},"name":{"type":"string","example":"doggie"},"photoUrls":{"type":"array","xml":{"wrapped":true},"items":{"type":"string","xml":{"name":"photoUrl"}}},"tags":{"type":"array","xml":{"wrapped":true},"items":{"xml":{"name":"tag"},"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}}}},"status":{"type":"string","description":"pet status in the store","enum":["available","pending","sold"]}},"xml":{"name":"Pet"},"$id":"RequestBody","$schema":"http://json-schema.org/draft-07/schema#"},
      "type": "json",
    },
    "responseBody": Object {
      "all": {"$id":"Response","oneOf":[{"$ref":"Response405"}],"$schema":"http://json-schema.org/draft-07/schema#"},
      "statuses": Object {
        "405": {"$id":"Response405","type":"boolean","not":true,"$schema":"http://json-schema.org/draft-07/schema#"},
      },
      "success": null,
    },
  },
}
`;

exports[`getByMethodAndPath /v2/petstore.json 3`] = `
Object {
  "info": {"operationId":"addPet","path":"/pet","method":"post"},
  "schema": Object {
    "headers": null,
    "params": null,
    "queries": null,
    "requestBody": Object {
      "contentType": "application/json",
      "schema": {"type":"object","required":["name","photoUrls"],"properties":{"id":{"type":"integer","format":"int64"},"category":{"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}},"xml":{"name":"Category"}},"name":{"type":"string","example":"doggie"},"photoUrls":{"type":"array","xml":{"wrapped":true},"items":{"type":"string","xml":{"name":"photoUrl"}}},"tags":{"type":"array","xml":{"wrapped":true},"items":{"xml":{"name":"tag"},"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}}}},"status":{"type":"string","description":"pet status in the store","enum":["available","pending","sold"]}},"xml":{"name":"Pet"},"$id":"RequestBody","$schema":"http://json-schema.org/draft-07/schema#"},
      "type": "json",
    },
    "responseBody": Object {
      "all": {"$id":"Response","oneOf":[{"$ref":"Response405"}],"$schema":"http://json-schema.org/draft-07/schema#"},
      "statuses": Object {
        "405": {"$id":"Response405","type":"boolean","not":true,"$schema":"http://json-schema.org/draft-07/schema#"},
      },
      "success": null,
    },
  },
}
`;

exports[`getByMethodAndPath /v3/petstore.json 1`] = `{"operationId":"addPet","method":"post","path":"/pet"}`;

exports[`getByMethodAndPath /v3/petstore.json 2`] = `
Object {
  "info": {"operationId":"addPet","path":"/pet","method":"post"},
  "schema": Object {
    "headers": null,
    "params": null,
    "queries": null,
    "requestBody": Object {
      "contentType": "application/json",
      "schema": {"required":["name","photoUrls"],"type":"object","properties":{"id":{"type":"integer","format":"int64","example":10},"name":{"type":"string","example":"doggie"},"category":{"type":"object","properties":{"id":{"type":"integer","format":"int64","example":1},"name":{"type":"string","example":"Dogs"}},"xml":{"name":"category"}},"photoUrls":{"type":"array","xml":{"wrapped":true},"items":{"type":"string","xml":{"name":"photoUrl"}}},"tags":{"type":"array","xml":{"wrapped":true},"items":{"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}},"xml":{"name":"tag"}}},"status":{"type":"string","description":"pet status in the store","enum":["available","pending","sold"]}},"xml":{"name":"pet"},"$id":"RequestBody","$schema":"http://json-schema.org/draft-07/schema#"},
      "type": "json",
    },
    "responseBody": Object {
      "all": {"$id":"Response","oneOf":[{"$ref":"Response200"},{"$ref":"Response405"}],"$schema":"http://json-schema.org/draft-07/schema#"},
      "statuses": Object {
        "200": {"required":["name","photoUrls"],"type":"object","properties":{"id":{"type":"integer","format":"int64","example":10},"name":{"type":"string","example":"doggie"},"category":{"type":"object","properties":{"id":{"type":"integer","format":"int64","example":1},"name":{"type":"string","example":"Dogs"}},"xml":{"name":"category"}},"photoUrls":{"type":"array","xml":{"wrapped":true},"items":{"type":"string","xml":{"name":"photoUrl"}}},"tags":{"type":"array","xml":{"wrapped":true},"items":{"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}},"xml":{"name":"tag"}}},"status":{"type":"string","description":"pet status in the store","enum":["available","pending","sold"]}},"xml":{"name":"pet"},"$id":"Response200","$schema":"http://json-schema.org/draft-07/schema#"},
        "405": {"$id":"Response405","type":"boolean","not":true,"$schema":"http://json-schema.org/draft-07/schema#"},
      },
      "success": {"$id":"ResponseOk","oneOf":[{"$ref":"Response200"}],"$schema":"http://json-schema.org/draft-07/schema#"},
    },
  },
}
`;

exports[`getByMethodAndPath /v3/petstore.json 3`] = `
Object {
  "info": {"operationId":"addPet","path":"/pet","method":"post"},
  "schema": Object {
    "headers": null,
    "params": null,
    "queries": null,
    "requestBody": Object {
      "contentType": "application/json",
      "schema": {"required":["name","photoUrls"],"type":"object","properties":{"id":{"type":"integer","format":"int64","example":10},"name":{"type":"string","example":"doggie"},"category":{"type":"object","properties":{"id":{"type":"integer","format":"int64","example":1},"name":{"type":"string","example":"Dogs"}},"xml":{"name":"category"}},"photoUrls":{"type":"array","xml":{"wrapped":true},"items":{"type":"string","xml":{"name":"photoUrl"}}},"tags":{"type":"array","xml":{"wrapped":true},"items":{"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}},"xml":{"name":"tag"}}},"status":{"type":"string","description":"pet status in the store","enum":["available","pending","sold"]}},"xml":{"name":"pet"},"$id":"RequestBody","$schema":"http://json-schema.org/draft-07/schema#"},
      "type": "json",
    },
    "responseBody": Object {
      "all": {"$id":"Response","oneOf":[{"$ref":"Response200"},{"$ref":"Response405"}],"$schema":"http://json-schema.org/draft-07/schema#"},
      "statuses": Object {
        "200": {"required":["name","photoUrls"],"type":"object","properties":{"id":{"type":"integer","format":"int64","example":10},"name":{"type":"string","example":"doggie"},"category":{"type":"object","properties":{"id":{"type":"integer","format":"int64","example":1},"name":{"type":"string","example":"Dogs"}},"xml":{"name":"category"}},"photoUrls":{"type":"array","xml":{"wrapped":true},"items":{"type":"string","xml":{"name":"photoUrl"}}},"tags":{"type":"array","xml":{"wrapped":true},"items":{"type":"object","properties":{"id":{"type":"integer","format":"int64"},"name":{"type":"string"}},"xml":{"name":"tag"}}},"status":{"type":"string","description":"pet status in the store","enum":["available","pending","sold"]}},"xml":{"name":"pet"},"$id":"Response200","$schema":"http://json-schema.org/draft-07/schema#"},
        "405": {"$id":"Response405","type":"boolean","not":true,"$schema":"http://json-schema.org/draft-07/schema#"},
      },
      "success": {"$id":"ResponseOk","oneOf":[{"$ref":"Response200"}],"$schema":"http://json-schema.org/draft-07/schema#"},
    },
  },
}
`;

exports[`getByOperationId /v2/petstore.json 1`] = `{"operationId":"addPet","method":"post","path":"/pet"}`;

exports[`getByOperationId /v3/petstore.json 1`] = `{"operationId":"addPet","method":"post","path":"/pet"}`;
